### PR TITLE
KAFKA-15729: Add KRaft support in GetOffsetShellTest

### PR DESCRIPTION
**Pull request by Owen-CH-Leung**
https://issues.apache.org/jira/browse/KAFKA-15729

As per [KAFKA-15729](https://issues.apache.org/jira/browse/KAFKA-15729), this PR adds Kraft support to `GetOffsetShellTest`

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation (including upgrade notes)

### CODE REVIEW

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java [hunk 1]
```diff
@@ -104,6 +104,11 @@ public Row(String name, int partition, Long timestamp) {
             this.timestamp = timestamp;
         }
 
+        @Override
+        public String toString() {
+            return "Row[name:" + name + ",partition:" + partition + ",timestamp:" + timestamp;
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: Is it used for debugging?

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java [hunk 2]
```diff
@@ -182,14 +187,19 @@ public void testTopicPartitionsArg() {
         setUp();
 
         List<Row> offsets = executeAndParse("--topic-partitions", "topic1:0,topic2:1,topic(3|4):2,__.*:3");
-        List<Row> expected = Arrays.asList(
-                new Row("__consumer_offsets", 3, 0L),
+        ArrayList<Row> expected = new ArrayList<>(
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: It seems we can keep using `List<Row>`, right?

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java [hunk 3]
```diff
@@ -338,6 +348,10 @@ private void assertExitCodeIsOne(String... args) {
     }
 
     private List<Row> expectedOffsetsWithInternal() {
+        if (cluster.isKRaftTest()) {
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: I prefer to make callers use `expectedTestTopicOffsets` instead of adding if-else here.

**Review by chia7712 — COMMENTED** [review]
@Owen-CH-Leung thanks for contribution. please take a look at following comments. thanks

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java [hunk 4]
```diff
@@ -182,14 +187,19 @@ public void testTopicPartitionsArg() {
...
+        ArrayList<Row> expected = new ArrayList<>(
+            Arrays.asList(
                 new Row("topic1", 0, 1L),
                 new Row("topic2", 1, 2L),
                 new Row("topic3", 2, 3L),
                 new Row("topic4", 2, 4L)
+            )
         );
 
+        if (!cluster.isKRaftTest()) {
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: I feel this test needs to verify the match pattern `__.*:3`, so we can do read for KRaft mode to create the internal topics.

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java (hunk 1 above)
Comment (Owen-CH-Leung) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: Yes you are right. Removed.

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java (hunk 2 above)
Comment (Owen-CH-Leung) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: Sure - I've changed to use `List<Row>`

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java (hunk 3 above)
Comment (Owen-CH-Leung) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: Sure - I've moved the evaluation of `cluster.isKRaftTest()` to the tests instead of putting it in the helper API

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java (hunk 4 above)
Comment (Owen-CH-Leung) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: Currently we didn't create any internal topic when running in kraft mode, so we can't verify the match pattern `__.*:3`. Shall we create an internal topic when running in Kraft mode ?

Comment (chia7712) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: > Shall we create an internal topic when running in Kraft mode ?

yep, we can create a consumer to read records before verifying the match pattern

Comment (Owen-CH-Leung) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: Sure. I've created Consumer to subscribe and poll from the topic for this test, followed by verifying the pattern

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java [hunk 5]
```diff
@@ -283,7 +318,7 @@ public void testTopicPartitionsArgWithInternalExcluded() {
         assertEquals(expected, offsets);
     }
 
-    @ClusterTest
+    @ClusterTest(clusterType = Type.ZK)
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: Could we create consumer to create internal topic for this test case?

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java [hunk 6]
```diff
@@ -338,6 +373,7 @@ private void assertExitCodeIsOne(String... args) {
     }
 
     private List<Row> expectedOffsetsWithInternal() {
+
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: could you please revert this unnecessary change?

**Review by chia7712 — COMMENTED** [review]
@Owen-CH-Leung thanks for updating PR. a couple of comments left. PTAL

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java (hunk 5 above)
Comment (Owen-CH-Leung) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: Sure - amended the test to create consumer for polling

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java (hunk 6 above)
Comment (Owen-CH-Leung) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: Sure & reverted

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java [hunk 7]
```diff
@@ -47,14 +51,16 @@
...
-@ClusterTestDefaults(clusterType = Type.ZK)
+@ClusterTestDefaults(clusterType = Type.ALL)
 @Tag("integration")
 public class GetOffsetShellTest {
     private final int topicCount = 4;
     private final int offsetTopicPartitionCount = 4;
     private final ClusterInstance cluster;
     private final String topicName = "topic";
 
+    private final int consumerTimeout = 100;
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: How about using `Duration` here?

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java [hunk 8]
```diff
@@ -181,13 +214,17 @@ public void testTopicPatternArgWithPartitionsArg() {
     public void testTopicPartitionsArg() {
         setUp();
 
+        createConsumerAndPoll();
+
         List<Row> offsets = executeAndParse("--topic-partitions", "topic1:0,topic2:1,topic(3|4):2,__.*:3");
-        List<Row> expected = Arrays.asList(
+        List<Row> expected = new ArrayList<>(
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: This change is unnecessary now.

**Review by chia7712 — COMMENTED** [review]
@Owen-CH-Leung two trivial comments left. overall +1, and thanks for updating patch continually.

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java (hunk 7 above)
Comment (Owen-CH-Leung) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: Sure - Updated to use `Duration`

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java (hunk 8 above)
Comment (Owen-CH-Leung) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: Thanks. Reverted

**Review by chia7712 — APPROVED** [review]
LGTM

**Comment by chia7712** [comment]
@Owen-CH-Leung Could you please rebase code to trigger QA again?

**Comment by chia7712** [comment]
It seems all builds failed due to `tools` module. Could you check them please?
```
[2024-03-26T11:58:09.997Z] Execution failed for task ':tools:test'.

[2024-03-26T11:58:09.997Z] > Process 'Gradle Test Executor 84' finished with non-zero exit value 1

[2024-03-26T11:58:09.997Z]   This problem might be caused by incorrect test process configuration.

[2024-03-26T11:58:09.997Z]   For more on test execution, please refer to https://docs.gradle.org/8.6/userguide/java_testing.html#sec:test_execution in the Gradle documentation.
```

**Comment by chia7712** [comment]
@Owen-CH-Leung The `printUsageAndExit` call `exit(1)` so `KRAFT` and `CO_KRAFT` will stop the JVM. `ZK` can capture the exit code to throw exception so it does not terminate the JVM.

Hence, a simple solution is - set dumb exit procedure for `testPrintHelp`. For example:
```java
    @ClusterTest
    public void testPrintHelp() {
        Exit.setExitProcedure((statusCode, message) -> { });
        try {
            String out = ToolsTestUtils.captureStandardErr(() -> GetOffsetShell.mainNoExit("--help"));
            assertTrue(out.startsWith(GetOffsetShell.USAGE_TEXT));
        } finally {
            Exit.resetExitProcedure();
        }
    }
``` 
WDYT?

**Comment by Owen-CH-Leung** [comment]
> printUsageAndExit

Agree. Setting a dumb exit procedure solves the failed build. Let me revise that

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java [hunk 9]
```diff
@@ -48,13 +52,14 @@
...
 @ExtendWith(value = ClusterTestExtensions.class)
-@ClusterTestDefaults(clusterType = Type.ZK)
+@ClusterTestDefaults(clusterType = Type.ALL)
 @Tag("integration")
 public class GetOffsetShellTest {
     private final int topicCount = 4;
     private final int offsetTopicPartitionCount = 4;
     private final ClusterInstance cluster;
     private final String topicName = "topic";
+    private final Duration consumerTimeout = Duration.ofMillis(100);
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: Could you increase the timeout to run QA again? The failed tests are caused that `__consumer_offsets` not created.

Comment (Owen-CH-Leung) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: Sure. Bumped the timeout to 1000.

**Comment by chia7712** [comment]
@Owen-CH-Leung Could you add `toString` back? It seems the test is unstable and we need to dig in it :)

**Comment by Owen-CH-Leung** [comment]
No prob. Added back `toString` to troubleshoot

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java (hunk 1 above)
Comment (chia7712) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: Could you rename the `timestamp` to `offset`?

Comment (Owen-CH-Leung) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: you mean like `return "Row[name:" + name + ",partition:" + partition + ",offset:" + offset;` ?

Comment (chia7712) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: yep, the output from shell is offset rather than timestamp

Comment (Owen-CH-Leung) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: I tried to do the refactoring to use `offset` rather than `timestamp`, but I'm a bit lost as in how I can get the offset number from the `GetOffsetShell`. I tried to read through the CI output but couldn't locate  where the output is an offset rather than the timestamp. Could you point me to the outputs that you saw so that I can get more context of it ? Thanks

Comment (chia7712) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: https://github.com/apache/kafka/blob/trunk/tools/src/main/java/org/apache/kafka/tools/GetOffsetShell.java#L93

Comment (Owen-CH-Leung) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: Thanks. Amended to use offset instead of timestamp

**Comment by chia7712** [comment]
rebase code to include the fix https://github.com/apache/kafka/commit/9a9b532d5d5beeecf7a4b769731ee609625429e1

**Comment by chia7712** [comment]
@Owen-CH-Leung Could you use `TestUtils.waitForCondition` to verify the records? maybe our QA is too slow to update the metadata before we do the check

**Comment by mcmmining** [comment]
> > printUsageAndExit
> 
> Agree. Setting a dumb exit procedure solves the failed build. Let me revise that

**Comment by Owen-CH-Leung** [comment]
> @Owen-CH-Leung Could you use `TestUtils.waitForCondition` to verify the records? maybe our QA is too slow to update the metadata before we do the check

Sure. I've used `waitForCondition` to wait for `__consumer_offsets` to be created when creating consumers to poll

**Comment by chia7712** [comment]
> Sure. I've used waitForCondition to wait for __consumer_offsets to be created when creating consumers to poll

I meant `assertEquals(expectedTestTopicOffsets().stream().filter(r -> r.partition <= 1).collect(Collectors.toList()), offsets);` should be retried since the metadata is not updated.

**Comment by Owen-CH-Leung** [comment]
> assertEquals(expectedTestTopicOffsets().stream().filter(r -> r.partition <= 1).collect(Collectors.toList()), offsets);

Oh got it. I read through the previous CI failure and notice that some tests like `testInternalExcluded`, `testNoFilterOptions` are intermittently failing also. Shall we just add the retry logic for tests that perform assertion based on the Row class ?

**Comment by chia7712** [comment]
> Shall we just add the retry logic for tests that perform assertion based on the Row class ?

yep, `waitForCondition` can address that for you :)

**Comment by Owen-CH-Leung** [comment]
> > Shall we just add the retry logic for tests that perform assertion based on the Row class ?
> 
> yep, `waitForCondition` can address that for you :)

Ok. I've implemented retry logic using `waitForCondition`. Let's see how the build goes.

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java [hunk 10]
```diff
@@ -391,4 +447,15 @@ private String[] addBootstrapServer(String... args) {
 
         return newArgs.toArray(new String[0]);
     }
+
+    private void retryUntilEqual(List<Row> expected, List<Row> output) {
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: the `output` should be a function. Otherwise, the output always returns same value in the waiting loop

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java [hunk 11]
```diff
@@ -94,15 +101,47 @@ private void setUp() {
...
+        props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+
+        try (KafkaConsumer<String, String> consumer = new KafkaConsumer<>(props)) {
+            List<String> topics = new ArrayList<>();
+            for (int i = 0; i < topicCount + 1; i++) {
+                topics.add(getTopicName(i));
+            }
+            consumer.subscribe(topics);
+            consumer.poll(consumerTimeout);
+            TestUtils.waitForCondition(
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: we can get rid of this wait

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java (hunk 10 above)
Comment (Owen-CH-Leung) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: Ahh yes. I've revised to accept a function now. The output should be refreshed during the wait loop. Let's see how the build goes

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java (hunk 11 above)
Comment (Owen-CH-Leung) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: Yess. Removed

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java (hunk 10 above)
Comment (Owen-CH-Leung) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: @chia7712 The build still contains falied test even when retry is implemented... looks like it's due to race condition when executing the tests ?

Comment (chia7712) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: > The build still contains falied test even when retry is implemented... looks like it's due to race condition when executing the tests ?

all failed tests are running in kraft mode, so I agree up to a point...

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java [hunk 12]
```diff
@@ -391,4 +420,15 @@ private String[] addBootstrapServer(String... args) {
 
         return newArgs.toArray(new String[0]);
     }
+
+    private void retryUntilEqual(List<Row> expected, Supplier<List<Row>> outputSupplier) {
+        try {
+            TestUtils.waitForCondition(
+                    () -> expected.equals(outputSupplier.get()),
+                    "TopicOffsets did not match. Expected: " + expectedTestTopicOffsets() + ", but was: " + outputSupplier.get()
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: Could you use `admin` to list latest offsets after it fails? for example:

```java
    private Set<Row> parse() throws ExecutionException, InterruptedException {
        try (Admin admin = cluster.createAdminClient()) {
            Set<String> topics = admin.listTopics(new ListTopicsOptions().listInternal(true)).listings().get()
                    .stream().map(TopicListing::name).collect(Collectors.toSet());
            Map<TopicPartition, OffsetSpec> offsetRequest = admin.describeTopics(topics)
                    .allTopicNames().get().entrySet().stream().flatMap(entry -> entry.getValue().partitions()
                            .stream().map(p -> new AbstractMap.SimpleImmutableEntry<>(new TopicPartition(entry.getKey(), p.partition()), OffsetSpec.latest())))
                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
            return admin.listOffsets(offsetRequest).all().get().entrySet().stream()
                    .map(entry -> new Row(entry.getKey().topic(), entry.getKey().partition(), entry.getValue().offset()))
                    .collect(Collectors.toSet());
        }
    }
```

Comment (Owen-CH-Leung) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: So sth like below ? Print out the final offset as error msg for debugging ?
```
TestUtils.waitForCondition(
                    () -> expected.equals(outputSupplier.get()),
                    "TopicOffsets did not match. Expected: " + expectedTestTopicOffsets() + ", but was: "
                            + outputSupplier.get() + ". Final offsets: " + getFinalOffsets()
            );
```

Comment (chia7712) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: yep, I loop the test with this PR 1000 times, all pass :(

It seems we need to reproduce the failure on our CI

Comment (Owen-CH-Leung) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: Ok done. Let's see what we get from there =)

Comment (chia7712) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: I have the similar error in another PR (https://github.com/apache/kafka/pull/15621) :(

**Comment by chia7712** [comment]
blocked by #15663

**Comment by chia7712** [comment]
rebase to trigger QA again

**Comment by Owen-CH-Leung** [comment]
> rebase to trigger QA again

Look like the build still contains failed test :(

**Comment by chia7712** [comment]
> Look like the build still contains failed test :(

yep, I have filed another #15654 to dig in that :_

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java [hunk 13]
```diff
@@ -94,15 +109,38 @@ private void setUp() {
...
+            consumer.subscribe(topics);
+            consumer.poll(consumerTimeout);
+        }
+    }
+
     static class Row {
         private String name;
         private int partition;
-        private Long timestamp;
+        private Long offset;
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: Could you add `final` to those fields?

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java [hunk 14]
```diff
@@ -48,13 +62,14 @@
...
 @ExtendWith(value = ClusterTestExtensions.class)
-@ClusterTestDefaults(clusterType = Type.ZK)
+@ClusterTestDefaults(clusterType = Type.ALL)
 @Tag("integration")
 public class GetOffsetShellTest {
     private final int topicCount = 4;
     private final int offsetTopicPartitionCount = 4;
     private final ClusterInstance cluster;
     private final String topicName = "topic";
+    private final Duration consumerTimeout = Duration.ofMillis(1000);
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: this can be a local variable

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java [hunk 15]
```diff
@@ -391,4 +428,30 @@ private String[] addBootstrapServer(String... args) {
...
+                    () -> expected.equals(outputSupplier.get()),
+                    "TopicOffsets did not match. Expected: " + expectedTestTopicOffsets() + ", but was: "
+                            + outputSupplier.get() + ". Final offsets: " + getFinalOffsets()
+            );
+        } catch (InterruptedException | ExecutionException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private Set<Row> getFinalOffsets() throws ExecutionException, InterruptedException {
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: we can remove this function now.

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java [hunk 16]
```diff
@@ -391,4 +428,30 @@ private String[] addBootstrapServer(String... args) {
 
         return newArgs.toArray(new String[0]);
     }
+
+    private void retryUntilEqual(List<Row> expected, Supplier<List<Row>> outputSupplier) {
+        try {
+            TestUtils.waitForCondition(
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: we don't need this retry now

**Review by chia7712 — COMMENTED** [review]
@Owen-CH-Leung 

The root cause is that not all produce records succeed to be sent, and we don't check all sends before closing producer. As our CI is very busy, it could case timeout if we send many records at once. Hence, could you call `get` on the `Producer#send` to make sure those records get sent.

**Comment by Owen-CH-Leung** [comment]
> @Owen-CH-Leung
> 
> The root cause is that not all produce records succeed to be sent, and we don't check all sends before closing producer. As our CI is very busy, it could case timeout if we send many records at once. Hence, could you call `get` on the `Producer#send` to make sure those records get sent.

Thanks a lot for helping to identify the root cause! I've just pushed a commit to address all your comments. Let's see what we get from the build =)

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java [hunk 17]
```diff
@@ -48,7 +53,7 @@
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
 @ExtendWith(value = ClusterTestExtensions.class)
-@ClusterTestDefaults(clusterType = Type.ZK)
+@ClusterTestDefaults(clusterType = Type.ALL)
 @Tag("integration")
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java:61]: Could you use `serverProperties` to define default server configs?
```java
        cluster.config().serverProperties().put("auto.create.topics.enable", false);
        cluster.config().serverProperties().put("offsets.topic.replication.factor", "1");
        cluster.config().serverProperties().put("offsets.topic.num.partitions", String.valueOf(offsetTopicPartitionCount));
```
could be replaced by 
```java
@ClusterTestDefaults(clusterType = Type.ALL, serverProperties = {
    @ClusterConfigProperty(key = "auto.create.topics.enable", value = "false"),
    @ClusterConfigProperty(key = "offsets.topic.replication.factor", value = "1"),
    @ClusterConfigProperty(key = "offsets.topic.num.partitions", value = "4")
})
```

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java [hunk 18]
```diff
@@ -48,7 +53,7 @@
...
 public class GetOffsetShellTest {
     private final int topicCount = 4;
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java:63]: `topicName` could be a local variable

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java [hunk 19]
```diff
@@ -81,28 +86,57 @@ private void setUp() {
...
         props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
         props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
 
         try (KafkaProducer<String, String> producer = new KafkaProducer<>(props)) {
             IntStream.range(0, topicCount + 1)
                 .forEach(i -> IntStream.range(0, i * i)
-                        .forEach(msgCount -> producer.send(
-                                new ProducerRecord<>(getTopicName(i), msgCount % i, null, "val" + msgCount)))
+                        .forEach(msgCount -> {
+                            try {
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: we can leverage `assertDoesNotThrow` to simplify the code
```java
assertDoesNotThrow(() -> producer.send(
                                    new ProducerRecord<>(getTopicName(i), msgCount % i, null, "val" + msgCount)).get());
```

**Review by chia7712 — COMMENTED** [review]
@Owen-CH-Leung thanks for nice updates. the flaky is gone. please take a look at refactor-related comments. thanks!

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java (hunk 17 above)
Comment (Owen-CH-Leung) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java:61]: those server config are annotated with `@BeforeEach`. Do you mean you want to put them like below ?

```
@ClusterTest(clusterType = Type.ALL, serverProperties = {
            @ClusterConfigProperty(key = "auto.create.topics.enable", value = "false"),
            @ClusterConfigProperty(key = "offsets.topic.replication.factor", value = "1"),
            @ClusterConfigProperty(key = "offsets.topic.num.partitions", value = "4")
    })
```

If so we'd have to modify each test. `ClusterTestDefaults` didn't have the API `serverProperties` so I think it wouldn't compile

Comment (chia7712) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java:61]: > ClusterTestDefaults didn't have the API serverProperties so I think it wouldn't compile

It is supported now. see #15687

Comment (Owen-CH-Leung) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java:61]: Thanks! I've pushed a commit to address all your comments.

**Comment by chia7712** [comment]
failed tests pass on my local. will merge it.
```
./gradlew cleanTest :streams:test --tests SlidingWindowedKStreamIntegrationTest.shouldRestoreAfterJoinRestart --tests StreamsAssignmentScaleTest.testHighAvailabilityTaskAssignorLargeNumConsumers :tools:test --tests MetadataQuorumCommandTest.testDescribeQuorumReplicationSuccessful --tests MetadataQuorumCommandTest.testDescribeQuorumStatusSuccessful --tests ReassignPartitionsIntegrationTest.testProduceAndConsumeWithReassignmentInProgress --tests ReassignPartitionsIntegrationTest.testReassignment --tests ReassignPartitionsIntegrationTest.testHighWaterMarkAfterPartitionReassignment :storage:test --tests TransactionsWithTieredStoreTest.testAbortTransactionTimeout :connect:runtime:test --tests org.apache.kafka.connect.integration.RebalanceSourceConnectorsIntegrationTest.testAddingWorker --tests org.apache.kafka.connect.integration.RebalanceSourceConnectorsIntegrationTest.testRemovingWorker :trogdor:test --tests CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated :connect:mirror:test --tests MirrorConnectorsIntegrationSSLTest.testSyncTopicConfigs :core:test --tests DelegationTokenEndToEndAuthorizationWithOwnerTest.testProduceConsumeViaSubscribe --tests DelegationTokenEndToEndAuthorizationWithOwnerTest.testCreateUserWithDelegationToken --tests ConsumerBounceTest.testConsumptionWithBrokerFailures --tests ConsumerBounceTest.testSeekAndCommitWithBrokerFailures --tests PlaintextConsumerTest.testCoordinatorFailover --tests SaslMultiMechanismConsumerTest.testCoordinatorFailover --tests SslConsumerTest.testCoordinatorFailover
```

**Comment by chia7712** [comment]
@Owen-CH-Leung thanks for your contribution and effort!
